packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,311 @@
+                                <pl-vault-item-list-item
+                                    .item=${listItem.item}
+                                    .vault=${listItem.vault}
+                                    class="stretch collapse"
+                                ></pl-vault-item-list-item>
+                            </div>
+                        `
+                    )}
+                </pl-list>
+                <pl-button
+                    class="slim margined transparent"
+                    @click=${() => this.go("items", routeParams)}
+                    ?hidden=${listItems.length < 6}
+                >
+                    <div>${$l("Show All")}</div>
+                    <pl-icon icon="arrow-right" class="small left-margined"></pl-icon>
+                </pl-button>
+            </section>
+        `;
+    }
+}
+
+async function sha1(password: string) {
+    const hashedPasswordData = await getCryptoProvider().hash(
+        stringToBytes(password),
+        new HashParams({ algorithm: "SHA-1" })
+    );
+    const hashedPassword = bytesToHex(hashedPasswordData);
+    return hashedPassword;
+}
+
+async function isPasswordWeak(password: string) {
+    const { score } = await passwordStrength(password);
+
+    return score < 2;
+}
+
+async function hasPasswordBeenCompromised(passwordHash: string) {
+    const hashPrefix = passwordHash.slice(0, 5);
+
+    const response = await fetch(`https://api.pwnedpasswords.com/range/${hashPrefix}`);
+
+    const result = await response.text();
+
+    const matchingHashSuffixes = result.split("\r\n");
+
+    for (const matchingHashSuffix of matchingHashSuffixes) {
+        const fullLowercaseHash = `${hashPrefix}${matchingHashSuffix.toLowerCase().split(":")[0]}`;
+
+        if (fullLowercaseHash === passwordHash) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+export async function auditVaults(
+    vaults: Vault[],
+    {
+        updateOnlyItemWithId,
+        updateOnlyIfOutdated,
+    }: { updateOnlyItemWithId?: string; updateOnlyIfOutdated?: boolean } = {}
+) {
+    const reusedPasswords: ListItem[] = [];
+    const weakPasswords: ListItem[] = [];
+    const compromisedPasswords: ListItem[] = [];
+
+    const firstMatchPerPasswordHash = new Map<
+        string,
+        { item: VaultItem; vault: Vault; passwordField: { field: Field; fieldIndex: number } }
+    >();
+    const reusedPasswordItemIds: Set<string> = new Set();
+    const weakPasswordItemIds: Set<string> = new Set();
+    const compromisedPasswordItemIds: Set<string> = new Set();
+
+    const oneWeekAgo = sub(new Date(), { weeks: 1 });
+
+    // TODO: Remove this
+    console.log(`Running audit! ${JSON.stringify({ updateOnlyItemWithId, updateOnlyIfOutdated })}`);
+
+    for (const vault of vaults) {
+        for (const item of vault.items) {
+            if (updateOnlyItemWithId) {
```

Comment:
I think you can get rid of that check and parameter, since we can always ignore items without an id.

Reply:
This is a check to make sure we can run the audit for a single item (without refactoring the code a bunch).

Reply:
Ah, I thought this was a boolean to filter out items with no ids at all 🤦‍♂️. Why not  `if (updateOnlyItemWithId && item.id !== updateOnlyItemWithId)` though, instead of two nested conditionals?

Reply:
It could be that and I had that initially, but it felt harder to read than this. I can change it if you want.